### PR TITLE
ATM: Extract training data

### PR DESCRIPTION
**Pull request by tiferet**
Implement the new query that selects data for training. 

For now we include clauses that implement logic that is identical to the old queries, so that the final dataset is identical, but we also note as TODO items some constraints we may want to experiment with removing. We include a temporary wrapper query that converts the resulting data into the format expected by the endpoint pipeline.

This PR moves the couple of small pieces of `ExtractEndpointData` that are still needed into `ExtractEndpointDataTraining.qll`, and deletes `ExtractEndpointDataTraining.ql`, `ExtractEndpointDataTraining.qll`, and the associated test files.

Timing checks:
- ✅ KPI timing experiment: https://github.com/github/codeql-dca-main/issues/8563
- ☑️ The local runtime of `endpoint_large_scale/ExtractEndpointDataTraining` is slightly impacted: Increased from about 4s to about 5s.

Closes https://github.com/github/ml-ql-adaptive-threat-modeling/issues/2098

### CODE REVIEW

**Comment by tiferet** [comment]
If the tests pass (they've been running for almost an hour now) and the [KPI experiment](https://github.com/github/codeql-dca-main/issues/8563) is OK, then this PR will be ready for review when you get to work on Tuesday.

**Comment by tiferet** [comment]
Nice-- the checks finally passed ✅

**Comment by tiferet** [comment]
@kaeluka The KPI timing experiment is OK, so this PR is now ready for review 🏓

**Review by kaeluka — DISMISSED** [review]
👍 this looks good to me. I've left a few *nitpicks* (in other words: not necessary to address before merging — feel free to address in your next PR).

I've tried to use the suggestions feature where possible to make this as effortless as possible on your part.

The one choice you should make before merging is what to do about my suggestion to delete the logic backing up the `hasFlowFromSource` value. If you accept the suggestion, you'd need to also fix the .expected file. You may ignore that suggestion, merge and I'll send a PR tomorrow that removes the flag after you merge this PR.

File: javascript/ql/experimental/adaptivethreatmodeling/modelbuilding/extraction/ExtractEndpointDataTraining.qll [hunk 1]
```diff
@@ -0,0 +1,239 @@
...
+import experimental.adaptivethreatmodeling.XssATM as XssAtm
+
+/**
+ * Gets the set of featureName-featureValue pairs for each endpoint in the training set.
+ *
+ * `EndpointFeatures::tokenFeatures` has no results when `featureName` is absent for the endpoint
+ * `endpoint`. To preserve compatibility with the data pipeline, this relation will instead set
+ * `featureValue` to the empty string in this case.
+ */
+query predicate tokenFeatures(DataFlow::Node endpoint, string featureName, string featureValue) {
```
[the remark this answers is not in the document]
Comment (tiferet) [reply javascript/ql/experimental/adaptivethreatmodeling/modelbuilding/extraction/ExtractEndpointDataTraining.qll:25]: OoI, what's the advantage of deleting the `query `? I read in the docs that it's good to put in `query predicate` because it allows us to run this query on a DB in the IDE for debugging purposes, which I've sometimes done. Is there a disadvantage to labeling something as a `query predicate`? Does that slow the query time, or something?

File: javascript/ql/experimental/adaptivethreatmodeling/modelbuilding/extraction/ExtractEndpointDataTraining.qll [hunk 2]
```diff
@@ -0,0 +1,239 @@
...
+ */
+query predicate trainingEndpoints(
+  DataFlow::Node endpoint, EndpointType endpointClass, EndpointCharacteristic characteristic
+) {
+  characteristic.getEndpoints(endpoint) and
+  // Only consider the source code for the project being analyzed.
+  exists(endpoint.getFile().getRelativePath()) and
+  // Only select endpoints that can be part of a tainted flow: Constant expressions always evaluate to a constant
+  // primitive value. Therefore they can't ever appear in an alert, making them less interesting training examples.
+  // TODO: Experiment with removing this requirement.
```
[the remark this answers is not in the document]
Comment (tiferet) [reply javascript/ql/experimental/adaptivethreatmodeling/modelbuilding/extraction/ExtractEndpointDataTraining.qll:62]: Hmm, I suppose we could. Would that characteristic have any (positive or negative) implications for any class? Or are you suggesting adding it as a characteristic with no implications, simply so that the modeling code could use it in type balancing? 

Either way it would need to be done as an experiment, because it could impact ATM metrics, so we'd have to check whether it helps or hurts them.

File: javascript/ql/experimental/adaptivethreatmodeling/modelbuilding/extraction/ExtractEndpointDataTraining.qll [hunk 3]
```diff
@@ -0,0 +1,239 @@
...
+  not endpoint.asExpr() instanceof ConstantExpr and
+  // Do not select endpoints filtered out by end-to-end evaluation.
+  // TODO: Experiment with removing this requirement.
```
[the remark this answers is not in the document]
Comment (tiferet) [reply javascript/ql/experimental/adaptivethreatmodeling/modelbuilding/extraction/ExtractEndpointDataTraining.qll:65]: Same question as above https://github.com/github/codeql/pull/11263#discussion_r1023029280

File: javascript/ql/experimental/adaptivethreatmodeling/modelbuilding/extraction/ExtractEndpointDataTraining.qll [hunk 4]
```diff
@@ -0,0 +1,239 @@
...
+  not Exclusions::isFileExcluded(endpoint.getFile()) and
+  // Filter out negative examples that also have a LikelyNotASinkReason, because this is currently done here
+  // https://github.com/github/codeql/blob/387e57546bf7352f7c1cfe781daa1a3799b7063e/javascript/ql/experimental/adaptivethreatmodeling/modelbuilding/extraction/ExtractEndpointData.qll#L77
+  // TODO: Experiment with removing this requirement.
+  not (
+    endpointClass instanceof NegativeType and
+    exists(EndpointCharacteristic c |
+      c.getEndpoints(endpoint) and
+      c instanceof LikelyNotASinkCharacteristic
```
[the remark this answers is not in the document]
Comment (tiferet) [reply javascript/ql/experimental/adaptivethreatmodeling/modelbuilding/extraction/ExtractEndpointDataTraining.qll:74]: I agree, and in fact I'm pretty sure this entire clause needs to be removed. It doesn't make logical sense to me: If an endpoint has both a high-confidence reason to think it's not a sink _and_ a medium-confidence reason to think it's not as sink, why would we exclude it from the training set? (See the last item [here](https://github.com/github/ml-ql-adaptive-threat-modeling/issues/2058#issuecomment-1274691810)). I've put it in this PR because we want to reproduce the old data for now. Changing this logic needs to be tested end-to-end, because it might affect our metrics. After this PR is merged, we should start running experiments getting rid of the clauses we don't like and seeing if metrics improve (or at least stay the same).

File: javascript/ql/experimental/adaptivethreatmodeling/modelbuilding/extraction/ExtractEndpointDataTraining.qll [hunk 5]
```diff
@@ -0,0 +1,239 @@
...
+  query instanceof NosqlInjectionQuery and result instanceof NosqlInjectionAtm::Configuration
+  or
+  query instanceof SqlInjectionQuery and result instanceof SqlInjectionAtm::Configuration
+  or
+  query instanceof TaintedPathQuery and result instanceof TaintedPathAtm::Configuration
+  or
+  query instanceof XssQuery and result instanceof XssAtm::Configuration
+}
+
+// TODO: Delete this once we are no longer surfacing `hasFlowFromSource`.
```
[the remark this answers is not in the document]
Comment (tiferet) [reply javascript/ql/experimental/adaptivethreatmodeling/modelbuilding/extraction/ExtractEndpointDataTraining.qll:211]: The endpoint pipeline passes this value on to the modeling code. IIRC, in the modeling code I experimented with using it to weigh samples, but concluded that this added no benefit. I'd have to dig into the modeling code and make sure it doesn't use `hasFlowFromSource` anywhere, and then I could do that.

**Comment by tiferet** [comment]
Meta comment: In all the cases above, my goal in this PR was to reproduce the current data exactly, so that we won't need to do end-to-end testing before merging this PR. Once this basic framework is merged, the next step would be to open a PR that adjusts the logic in all the ways we _think_ it should be adjusted, then run end-to-end testing, and if metrics aren't hurt we could merge that PR and run a partial update process. Until the orchestrator is fully functional, that's a non-trivial process, so I'd rather not do it many times unnecessarily. That's why I prefer a strict separation between PRs that change the framework but leave the data _identical_, and can be verified with PR checks, followed by a small number of targeted PRs that improve the underlying logic/data but require end-to-end testing.

File: javascript/ql/experimental/adaptivethreatmodeling/modelbuilding/extraction/ExtractEndpointDataTraining.qll [hunk 6]
```diff
@@ -0,0 +1,239 @@
...
+ * Holds if the given endpoint should be included in the training set as a sample belonging to endpointClass, and has
+ * the given characteristic. This query uses the endpoint characteristics to select and label endpoints for the training
+ * set, and provides a list of characteristics for each endpoint in the training set, which is used in the modeling
+ * code.
+ *
+ * Params:
+ * endpoint: The endpoint to include / exclude.
+ * endpointClass: The sink type. Each EndpointType has a predicate getEncoding, which specifies the classifier class
+ * for this sink type. Class 0 is the negative class (non-sink). Each positive int corresponds to a single sink type.
+ * This gives us the label for the endpoint in the training data.
```
[the remark this answers is not in the document]
Comment (tiferet) [reply javascript/ql/experimental/adaptivethreatmodeling/modelbuilding/extraction/ExtractEndpointDataTraining.qll]: I removed this and referenced that documentation for more details in fc078a47fdc257453cad808cd5608ce92bfd4f01

File: javascript/ql/experimental/adaptivethreatmodeling/modelbuilding/extraction/ExtractEndpointDataTraining.qll (hunk 5 above)
[the remark this answers is not in the document]
Comment (tiferet) [reply javascript/ql/experimental/adaptivethreatmodeling/modelbuilding/extraction/ExtractEndpointDataTraining.qll:210]: Incidentally, there's already [an issue](https://github.com/github/ml-ql-adaptive-threat-modeling/issues/2109) for this deprecation in the epic.

**Review by kaeluka — APPROVED** [review]
Still lgtm 👍

File: javascript/ql/experimental/adaptivethreatmodeling/modelbuilding/extraction/ExtractEndpointDataTraining.qll (hunk 5 above)
[the remark this answers is not in the document]
Comment (tiferet) [reply javascript/ql/experimental/adaptivethreatmodeling/modelbuilding/extraction/ExtractEndpointDataTraining.qll:210]: > Meta comment: In all the cases above, my goal in this PR was to reproduce the current data exactly, so that we won't need to do end-to-end testing before merging this PR. Once this basic framework is merged, the next step would be to open a PR that adjusts the logic in all the ways we _think_ it should be adjusted, then run end-to-end testing. If metrics aren't hurt we could merge that PR and run a partial update process. Until the orchestrator is fully functional, that's a non-trivial process, so I'd rather not do it many times unnecessarily. That's why I prefer a strict separation between PRs that change the framework but leave the data _identical_, and can be verified with PR checks, followed by a small number of targeted PRs that improve the underlying logic/data but require end-to-end testing.

I just opened [an issue](https://github.com/github/ml-ql-adaptive-threat-modeling/issues/2126) to track the followup experiments so we don't forget about them.

**Comment by tiferet** [comment]
> Still lgtm 👍

Thank you! I had some followup questions about some of your comments (e.g. https://github.com/github/codeql/pull/11263#discussion_r1023029280), so let's keep discussing them in this PR even after it's merged. When we reach a conclusion I'll implement it in a subsequent PR.
